core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -280,6 +286,210 @@ class ReplicaManager(val config: KafkaConfig,
 
   private var logDirFailureHandler: LogDirFailureHandler = null
 
+  // initially true when using a Raft-based metadata quorum, and may flip-flop thereafter;
+  // always false when using ZooKeeper
+  @volatile private var fenced: Boolean = zkClient.isEmpty
```

Comment:
I am a little confused about the relationships between `fenced`, `fencedPartitionStates`, and `allPartitions`. I thought the advantage of having a `Fenced` state is that we would not need a separate collection because we can represent a fenced partition inside `allPartitions`. On the other hand, if fencing is really a global state as indicated by the flag here, then perhaps we need neither the `Fenced` state nor `fencedPartitionStates`. We could change `getOnlinePartitionOrException` to check for `Online(partition) if !fenced`.

Basically my concern is whether we can have clearly enforced invariants about the internal state here.

Reply:
I think we can eliminate `fencedPartitionStates` and keep the same information inside the `HostedPartition.Fenced` instance -- it does seem to be a waste to have both.

Regarding the `fenced` variable, I was thinking about how we would deal with becoming fenced during normal operations, well after we had started up and recovered our logs.  In that situation, I did not intend this to represent a state such that every partition we lead or follow immediately transitions into the fenced state.  For example, while we are fenced I think it is conceivable that we could continue to successfully fetch from other brokers for partitions that we believe we follow and successfully respond to fetch requests for partitions that we believe we lead.  I intended this flag to indicate that any metadata changes related to partitions that we receive while in this state should not be relied upon as being the latest information and should be remembered for later (and potentially updated along the way) until we are no longer fenced and we know we have the right information.  Basically, like at startup, the state indicates when we need to catch up on the metadata log before acting on any partition metadata we might receive while we are catching up.

So, to summarize what I was thinking:

- At startup it is true that every partition is fenced until after we initially catch up and recover logs, then we start leading/following
- Later on, if we become fenced, we could potentially still lead/follow for a while, but any new partition metadata that shows up isn't relied upon as being correct -- such partitions transition to the fenced state and the metadata is remembered for when we catch up again -- then we start leading/following as appropriate.
- There is a case where if we are fenced for too long it is possible that we might have to abandon our metadata entirely and fetch a new snapshot.  This would happen if we are at metadata log offset X, but while fenced the offset X+1 became unavailable as an independently-fetchable offset and is only available as part of a snapshot.  In this case we have to do a "soft restart" of sorts, and we have some reconciliation to do on our internal state.  This reconciliation could be done in a number of ways, but one way would be to stop all leading/following, clear `allPartitions`, set the `fenced` flag, and only clear the `fenced` flag after we have rwsd the latest snapshot and caught up.